lib/rust/flexer/src/lib.rs

```diff
@@ -14,26 +14,241 @@
+            Some(ix)
+        } else {
+            None
+        }
+    }
+
+    /// Ends states until the specified `state` is reached, leaving the lexer in `state`.
+    ///
+    /// If `state` does not exist on the lexer's stack, then the lexer will be left in the root
+    /// state.
+    pub fn pop_states_until(&mut self, state:group::Identifier) -> Vec<group::Identifier> {
+        let non_opt_root_state_position =
+            self.state_stack.iter().positions(|elem| *elem == state).last().unwrap_or(0);
+        let range = (non_opt_root_state_position + 1)..self.state_stack.len();
+        self.state_stack.drain(range).collect()
+    }
+
+    /// Checks if the lexer is currently in the state described by `state`.
+    pub fn in_state(&mut self, state:group::Identifier) -> bool {
+        self.current_state() == state
+    }
+}
+
+// === Trait Impls ===
+
+impl<Definition,Output,Reader> Deref for Flexer<Definition,Output,Reader> {
+    type Target = Definition;
+    fn deref(&self) -> &Self::Target {
+        &self.definition
+    }
+}
+
+impl<Definition,Output,Reader> DerefMut for Flexer<Definition,Output,Reader> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.definition
+    }
+}
+
+
+
+// ===================
+// === StageStatus ===
+// ===================
+
+/// The result of executing a single step of the DFA.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum StageStatus {
+    /// The initial state of a lexer stage.
+    Initial,
+    /// The stage exits successfully, having consumed a complete token.
+    ExitSuccess,
+    /// The stage exits unsuccessfully.
+    ExitFail,
+    /// A single step of the DFA has executed successfully.
+    ExitFinished,
+    /// The lexer should continue, transitioning to the included state.
+    ContinueWith(usize)
+}
+
+impl StageStatus {
+    /// Checks if the lexer stage should continue.
+    pub fn should_continue(&self) -> bool {
+        self.continue_as().is_some()
+    }
+
+    /// Obtains the state to which the lexer should transition, iff the lexer should continue.
+    pub fn continue_as(&self) -> Option<usize> {
+        match self {
+            StageStatus::Initial           => Some(0),
+            StageStatus::ContinueWith(val) => Some(*val),
+            _                              => None
+        }
+    }
+}
+
+
+// === Trait Impls ===
+
+impl Default for StageStatus {
+    fn default() -> Self {
+        StageStatus::Initial
+    }
+}
+
+
+
+// ==============
+// === Result ===
+// ==============
+
+/// The result of executing the lexer on a given input.
+#[derive(Clone,Debug)]
+pub enum Result<T> {
+    /// The lexer succeeded, returning the contained token stream.
+    Success(Vec<T>),
+    /// The lexer succeeded on part of the input, returning the contained token stream.
+    Partial(Vec<T>),
+    /// The lexer failed on the input, returning any tokens it _did_ manage to consume.
+    Failure(Vec<T>)
```

Comment:
I believe that in old implementation we also were providing here information how far we went (not only the tokens so far). It doesn't really matter in context of Enso as here lexer should NEVER fail, but I feel the info is missing here.